scarb/src/core/resolver.rs

```diff
@@ -96,3 +100,104 @@ impl From<TargetKind> for DependencyEdge {
+    /// has gone wrong such as:
+    ///
+    /// * something got seriously corrupted,
+    /// * a "mirror" is not actually a mirror as some changes were made,
+    /// * a replacement source was not actually a replacement, some changes were made.
+    ///
+    /// In all of these cases, we want to report an error to indicate that something is awry.
+    /// Normal execution (esp. just using the default registry) should never run into this.
+    pub fn check_checksums(&self, lockfile: &Lockfile) -> Result<()> {
+        for package_lock in &lockfile.packages {
+            let (locked, source_id) = match (package_lock.checksum.as_ref(), package_lock.source) {
+                (None, None) => continue,
+                (Some(_), None) => {
+                    unreachable!(
+                        "Package lock entry `{n} v{v}` has `checksum` but no `source` field.",
+                        n = package_lock.name,
+                        v = package_lock.version
+                    );
+                }
+                (locked, Some(source_id)) => (locked, source_id),
+            };
+
+            let id = PackageId::new(
+                package_lock.name.clone(),
+                package_lock.version.clone(),
+                source_id,
+            );
+
+            let Some(actual) = self.summaries.get(&id).map(|s| s.checksum.as_ref()) else {
+                continue;
+            };
+
+            match (actual, locked) {
+                // If the checksums are the same, or both are not present, then we are good.
+                (Some(actual), Some(locked)) if actual == locked => {}
+                (None, None) => {}
+
+                // If the locked checksum was not calculated, and the current checksum is `Some`,
+                // it may indicate that a source was erroneously replaced or was replaced with
+                // something that desires stronger checksum guarantees than can be afforded
+                // elsewhere.
+                (Some(_), None) => {
+                    bail!(formatdoc! {"
+                        checksum for `{id}` was not previously calculated, but now it could be
+
+                        this could be indicative of a few possible situations:
+
+                            * the source `{source_id}` did not previously support checksums, \
+                              but was replaced with one that does
+                            * newer Scarb implementations know how to checksum this source, \
+                              but this older implementation does not
+                            * the lock file is corrupt
+                    "});
+                }
+
+                // If our checksum has not been calculated, then it could mean that future Scarb
+                // figured out how to do it or more realistically we were overridden with a source
+                // that does not have checksums.
```

Comment:
```suggestion
                // If our checksum has not been calculated, then it could mean that future Scarb
                // figured out how to do it or the source has been shadowed by with
                // a different one thanks to some unknown future logic.
```